crates/libs/core/src/sync/cancel.rs

```diff
@@ -95,40 +101,48 @@ where
         self_impl.consume_content()
     }
 
-    // TODO: send and comsume is expected to happend accross threads.
-    // Even though we use a oneshot channel to send the signal,
-    // it might be safer to add another memory barrier here.
+    /// Set the content for the ctx.
+    /// Marks the ctx as completed.
     fn set_content(&self, content: T) {
         let prev = self.content.replace(Some(content));
-        assert!(prev.is_none())
+        assert!(prev.is_none());
+        self.set_complete();
     }
 
-    // can only be called once after set content.
+    /// Consumes the content set by set_content().
+    /// can only be called once after set content.
     fn consume_content(&self) -> crate::Result<T> {
-        let opt = self.content.take();
-        match opt {
-            Some(x) => Ok(x),
-            None => {
-                if !self.IsCompleted().as_bool() {
-                    return Err(FabricErrorCode::FABRIC_E_OPERATION_NOT_COMPLETE.into());
-                }
+        match self.check_complete() {
+            true => Ok(self.content.take().expect("content is consumed twice.")),
+            false => {
                 if self.token.is_cancelled() {
                     Err(FabricErrorCode::E_ABORT.into())
                 } else {
-                    panic!("content is consumed twice.")
+                    Err(FabricErrorCode::FABRIC_E_OPERATION_NOT_COMPLETE.into())
                 }
             }
         }
     }
 
+    /// Set the ctx as completed. Requires the ctx content to be set. Makes
+    /// the content available for access from other threads using barrier.
     fn set_complete(&self) {
-        self.is_completed.set(true);
+        self.is_completed
+            .store(true, std::sync::atomic::Ordering::Release);
+    }
+
+    /// Checks ctx is completed.
+    /// Makes sure content sets by other threads is visible from this thread.
+    fn check_complete(&self) -> bool {
+        self.is_completed.load(std::sync::atomic::Ordering::Acquire)
     }
 }
 
 impl<T> IFabricAsyncOperationContext_Impl for BridgeContext3<T> {
     fn IsCompleted(&self) -> crate::BOOLEAN {
-        self.is_completed.get().into()
+        self.is_completed
+            .load(std::sync::atomic::Ordering::Relaxed)
```

Comment:
This should only load the value without affect cpu caching from other thread.
The other place we need Release and Acquire because we need to sync cpu cache for the ctx content not only the bool value.